File: packages/sdk/pact-ts/src/invariants/capability.ts
import { canonicalizeJson } from "./json.ts";
import { verifyEd25519Signature } from "./crypto.ts";
import { parseJsonText } from "./errors.ts";

export type CapabilityTimeStatus = "valid" | "not_yet_valid" | "expired";

export interface DelegationLink {
  capability_id: string;
  delegator: string;
  delegatee: string;
  attenuations?: unknown[];
  timestamp: number;
  signature: string;
}

export interface CapabilityToken {
  id: string;
  issuer: string;
  subject: string;
  scope: unknown;
  issued_at: number;
  expires_at: number;
  delegation_chain?: DelegationLink[];
  signature: string;
}

export interface CapabilityVerification {
  signature_valid: boolean;
  delegation_chain_valid: boolean;
  time_valid: boolean;
  time_status: CapabilityTimeStatus;
}

function delegationLinkBody(link: DelegationLink): Omit<DelegationLink, "signature"> {
  const { signature: _signature, ...body } = link;
  return body;
}

function verifyDelegationChain(chain: DelegationLink[], maxDelegationDepth?: number): boolean {
  if (maxDelegationDepth !== undefined && chain.length > maxDelegationDepth) {
    return false;
  }

  let previous: DelegationLink | undefined;
  for (const current of chain) {
    if (!verifyEd25519Signature(canonicalizeJson(delegationLinkBody(current)), current.delegator, current.signature)) {
      return false;
    }

    if (previous) {
      if (previous.delegatee !== current.delegator) {
        return false;
      }
      if (current.timestamp < previous.timestamp) {
        return false;
      }
    }
    previous = current;
  }

  return true;
}

export function parseCapabilityJson(input: string): CapabilityToken {
  return parseJsonText(input);
}

export function capabilityBody(capability: CapabilityToken): Omit<CapabilityToken, "signature"> {
  const { signature: _signature, ...body } = capability;
  return body;
}

export function capabilityBodyCanonicalJson(capability: CapabilityToken): string {
  return canonicalizeJson(capabilityBody(capability));
}

export function verifyCapability(
  capability: CapabilityToken,
  now: number,
  maxDelegationDepth?: number,
): CapabilityVerification {
  let time_status: CapabilityTimeStatus;
  if (now < capability.issued_at) {
    time_status = "not_yet_valid";
  } else if (now >= capability.expires_at) {
    time_status = "expired";
  } else {
    time_status = "valid";
  }

  return {
    signature_valid: verifyEd25519Signature(
      capabilityBodyCanonicalJson(capability),
      capability.issuer,
      capability.signature,
    ),
    delegation_chain_valid: verifyDelegationChain(capability.delegation_chain ?? [], maxDelegationDepth),
    time_valid: time_status === "valid",
    time_status,
  };
}

export function verifyCapabilityJson(
  input: string,
  now: number,
  maxDelegationDepth?: number,
): CapabilityVerification {
  return verifyCapability(parseCapabilityJson(input), now, maxDelegationDepth);
}

File: packages/sdk/arc-py/src/arc/invariants/capability.py
from __future__ import annotations

from typing import Any

from ..errors import parse_json_text
from .json import canonicalize_json
from .signing import verify_utf8_message_ed25519


def parse_capability_json(input_text: str) -> dict[str, Any]:
    return parse_json_text(input_text)


def _capability_body(capability: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in capability.items() if key != "signature"}


def _delegation_link_body(link: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in link.items() if key != "signature"}


def capability_body_canonical_json(capability: dict[str, Any]) -> str:
    return canonicalize_json(_capability_body(capability))


def _verify_delegation_chain(
    chain: list[dict[str, Any]],
    max_delegation_depth: int | None,
) -> bool:
    if max_delegation_depth is not None and len(chain) > max_delegation_depth:
        return False
    for index, current in enumerate(chain):
        if not verify_utf8_message_ed25519(
            canonicalize_json(_delegation_link_body(current)),
            current["delegator"],
            current["signature"],
        ):
            return False
        if index > 0:
            previous = chain[index - 1]
            if previous["delegatee"] != current["delegator"]:
                return False
            if current["timestamp"] < previous["timestamp"]:
                return False
    return True


def verify_capability(
    capability: dict[str, Any],
    now: int,
    max_delegation_depth: int | None = None,
) -> dict[str, Any]:
    if now < capability["issued_at"]:
        time_status = "not_yet_valid"
    elif now >= capability["expires_at"]:
        time_status = "expired"
    else:
        time_status = "valid"
    return {
        "signature_valid": verify_utf8_message_ed25519(
            capability_body_canonical_json(capability),
            capability["issuer"],
            capability["signature"],
        ),
        "delegation_chain_valid": _verify_delegation_chain(
            capability.get("delegation_chain", []),
            max_delegation_depth,
        ),
        "time_valid": time_status == "valid",
        "time_status": time_status,
    }


def verify_capability_json(
    input_text: str,
    now: int,
    max_delegation_depth: int | None = None,
) -> dict[str, Any]:
    return verify_capability(parse_capability_json(input_text), now, max_delegation_depth)
